(Partially) fix bug #49053: retina scaling of figures
https://savannah.gnu.org/bugs/?49053

This is changset http://hg.savannah.gnu.org/hgweb/octave/rev/4e108574385c
backported to release-4-4-1

diff -r 1f46d371968c libgui/graphics/Figure.cc
--- a/libgui/graphics/Figure.cc	Thu Aug 09 14:20:32 2018 -0400
+++ b/libgui/graphics/Figure.cc	Tue Sep 24 16:22:28 2019 +0200
@@ -41,6 +41,8 @@
 #include <QtDebug>
 #include <QTimer>
 #include <QToolBar>
+#include <QWindow>
+#include <QScreen>
 
 #include "Canvas.h"
 #include "Container.h"
@@ -133,6 +135,11 @@
 
     figure::properties& fp = properties<figure> ();
 
+    // Register for the signal that indicates when a window has moved
+    // to a different screen
+    connect (win, SIGNAL (figureWindowShown ()),
+             this, SLOT (figureWindowShown ()));
+     
     // Status bar
     m_statusBar = win->statusBar ();
     int boffset = 0;
@@ -1078,6 +1085,34 @@
       canvas->autoAxes (m_handle);
   }
 
+   void
+   Figure::figureWindowShown ()
+   {
+#if defined (HAVE_QSCREEN_DEVICEPIXELRATIO)
+      QWindow* window = qWidget<QMainWindow> ()->windowHandle ();
+      QScreen* screen = window->screen ();
+      
+      gh_manager::auto_lock lock;
+      figure::properties& fp = properties<figure> ();
+      fp.set___device_pixel_ratio__ (screen->devicePixelRatio ());
+      
+      connect (window, SIGNAL (screenChanged (QScreen*)),
+               +             this, SLOT (screenChanged (QScreen*)));
+#endif
+   }
+   
+   void
+   Figure::screenChanged (QScreen* screen)
+   {
+#if defined (HAVE_QSCREEN_DEVICEPIXELRATIO)
+      gh_manager::auto_lock lock;
+      
+      figure::properties& fp = properties<figure> ();
+      fp.set___device_pixel_ratio__ (screen->devicePixelRatio ());
+#endif
+   }
+   
+   
   void
   Figure::enableMouseTracking (void)
   {
diff -r 1f46d371968c libgui/graphics/Figure.h
--- a/libgui/graphics/Figure.h	Thu Aug 09 14:20:32 2018 -0400
+++ b/libgui/graphics/Figure.h	Tue Sep 24 16:22:28 2019 +0200
@@ -32,6 +32,7 @@
 
 class QMainWindow;
 class QToolBar;
+class QScreen;
 
 namespace QtHandles
 {
@@ -128,6 +129,8 @@
     void toggleAxes (void);
     void toggleGrid (void);
     void autoAxes (void);
+    void figureWindowShown ();
+    void screenChanged (QScreen*);
 
   public slots:
     uint8NDArray slotGetPixels (void);
diff -r 1f46d371968c libgui/graphics/FigureWindow.cc
--- a/libgui/graphics/FigureWindow.cc	Thu Aug 09 14:20:32 2018 -0400
+++ b/libgui/graphics/FigureWindow.cc	Tue Sep 24 16:22:28 2019 +0200
@@ -48,4 +48,9 @@
     return nullptr;
   }
 
+   void FigureWindow::showEvent (QShowEvent* ev)
+   {
+      QMainWindow::showEvent (ev);
+      emit figureWindowShown();
+   }
 }
diff -r 1f46d371968c libgui/graphics/FigureWindow.h
--- a/libgui/graphics/FigureWindow.h	Thu Aug 09 14:20:32 2018 -0400
+++ b/libgui/graphics/FigureWindow.h	Tue Sep 24 16:22:28 2019 +0200
@@ -37,10 +37,18 @@
     Q_OBJECT
 
   public:
-    FigureWindow (QWidget *parent = nullptr);
+    explicit FigureWindow (QWidget *parent = nullptr);
     ~FigureWindow (void);
 
     QMenu * createPopupMenu (void);
+     
+     
+  protected:
+     void showEvent(QShowEvent *ev);
+     
+  signals:
+     void figureWindowShown();
+     
   };
 
 }
diff -r 1f46d371968c libgui/graphics/GLCanvas.cc
--- a/libgui/graphics/GLCanvas.cc	Thu Aug 09 14:20:32 2018 -0400
+++ b/libgui/graphics/GLCanvas.cc	Tue Sep 24 16:22:28 2019 +0200
@@ -68,7 +68,9 @@
       {
         octave::opengl_renderer r;
 
-        r.set_viewport (width (), height ());
+        graphics_object fig = go.get_ancestor ("figure");
+        double dpr = fig.get ("__device_pixel_ratio__").double_value ();
+        r.set_viewport (dpr * width (), dpr * height ());
         r.draw (go);
       }
   }
@@ -171,11 +173,14 @@
 
     if (ax)
       {
-        octave::opengl_selector s;
+         octave::opengl_selector s;
 
-        s.set_viewport (width (), height ());
-        return s.select (ax, pt.x (), height () - pt.y (),
-                         octave::select_ignore_hittest);
+         graphics_object fig = ax.get_ancestor ("figure");
+         double dpr = fig.get ("__device_pixel_ratio__").double_value ();
+         
+         s.set_viewport (dpr * width (), dpr * height ());
+         return s.select (ax, pt.x (), height () - pt.y (),
+                          octave::select_ignore_hittest);
       }
 
     return graphics_object ();
diff -r 1f46d371968c libinterp/corefcn/gl-render.cc
--- a/libinterp/corefcn/gl-render.cc	Thu Aug 09 14:20:32 2018 -0400
+++ b/libinterp/corefcn/gl-render.cc	Tue Sep 24 16:22:28 2019 +0200
@@ -593,7 +593,8 @@
     : toolkit (), xform (), xmin (), xmax (), ymin (), ymax (),
       zmin (), zmax (), xZ1 (), xZ2 (), marker_id (), filled_marker_id (),
       camera_pos (), camera_dir (), interpreter ("none"), txt_renderer (),
-      selecting (false)
+      selecting (false),
+      m_devpixratio(1.)
   {
     // This constructor will fail if we don't have OpenGL or if the data
     // types we assumed in our public interface aren't compatible with the
@@ -694,6 +695,10 @@
   void
   opengl_renderer::draw_figure (const figure::properties& props)
   {
+     // Current physical-pixel to toolkit-pixel factor
+     m_devpixratio = props.get___device_pixel_ratio__ ();
+     
+     
     // Initialize OpenGL context
 
     init_gl_context (props.is_graphicssmoothing (), props.get_color_rgb ());
@@ -1123,6 +1128,11 @@
     glMultMatrixd (x_mat1.data ());
     glMatrixMode (GL_PROJECTION);
     glLoadIdentity ();
+     
+     // Use abstract Octave-pixels for transformation, not physical-pixels
+     vw[2] = octave::math::round (static_cast<float> (vw[2]) / m_devpixratio);
+     vw[3] = octave::math::round (static_cast<float> (vw[3]) / m_devpixratio);
+     
     glOrtho (0, vw[2], vw[3], 0, xZ1, xZ2);
     glMultMatrixd (x_mat2.data ());
     glMatrixMode (GL_MODELVIEW);
@@ -3528,6 +3538,10 @@
     glMatrixMode (GL_PROJECTION);
     glPushMatrix ();
     glLoadIdentity ();
+     
+     vp[2] = octave::math::round (static_cast<float> (vp[2]) / m_devpixratio);
+     vp[3] = octave::math::round (static_cast<float> (vp[3]) / m_devpixratio);
+     
     glOrtho (0, vp[2], vp[3], 0, xZ1, xZ2);
     glMatrixMode (GL_MODELVIEW);
     glPushMatrix ();
@@ -3700,7 +3714,8 @@
     if (i0 >= i1 || j0 >= j1)
       return;
 
-    glPixelZoom (pix_dx, -pix_dy);
+     glPixelZoom (m_devpixratio * pix_dx,
+                  - m_devpixratio * pix_dy);
     glRasterPos3d (im_xmin + nor_dx*j0, im_ymin + nor_dy*i0, 0);
 
     // by default this is 4
@@ -3915,7 +3930,8 @@
     txt_renderer.set_font (props.get ("fontname").string_value (),
                            props.get ("fontweight").string_value (),
                            props.get ("fontangle").string_value (),
-                           props.get ("__fontsize_points__").double_value ());
+                           props.get ("__fontsize_points__").double_value ()
+                           * m_devpixratio);
   }
 
   void
@@ -3953,7 +3969,7 @@
   {
 #if defined (HAVE_OPENGL)
 
-    glLineWidth (w);
+    glLineWidth (w * m_devpixratio);
 
 #else
 
@@ -4103,6 +4119,10 @@
     glMatrixMode (GL_PROJECTION);
     glPushMatrix ();
     glLoadIdentity ();
+     
+     vw[2] = octave::math::round (static_cast<float> (vw[2]) / m_devpixratio);
+     vw[3] = octave::math::round (static_cast<float> (vw[3]) / m_devpixratio);
+     
     glOrtho (0, vw[2], vw[3], 0, xZ1, xZ2);
     glMatrixMode (GL_MODELVIEW);
     glPushMatrix ();
diff -r 1f46d371968c libinterp/corefcn/gl-render.h
--- a/libinterp/corefcn/gl-render.h	Thu Aug 09 14:20:32 2018 -0400
+++ b/libinterp/corefcn/gl-render.h	Tue Sep 24 16:22:28 2019 +0200
@@ -223,7 +223,10 @@
     // Indicate we are drawing for selection purpose
     bool selecting;
   private:
-    class patch_tesselator;
+     
+     // Factor used for translating Octave pixels to actual device pixels
+     double m_devpixratio;
+     class patch_tesselator;
   };
 }
 
diff -r 1f46d371968c libinterp/corefcn/graphics.cc
--- a/libinterp/corefcn/graphics.cc	Thu Aug 09 14:20:32 2018 -0400
+++ b/libinterp/corefcn/graphics.cc	Tue Sep 24 16:22:28 2019 +0200
@@ -869,6 +869,20 @@
                            sz.extract_n (0, 2, 1, 2)).extract_n (0, 2, 1, 2);
 }
 
+static double
+device_pixel_ratio (graphics_handle h)
+{
+   double retval = 1.0;
+   
+   graphics_object fig = gh_manager::get_object (h).get_ancestor ("figure");
+   
+   if (fig.valid_object ())
+      retval = fig.get ("__device_pixel_ratio__").double_value ();
+   
+   return retval;
+}
+
+
 static void
 convert_cdata_2 (bool is_scaled, bool is_real, double clim_0, double clim_1,
                  const double *cmapv, double x, octave_idx_type lda,
@@ -2079,6 +2093,45 @@
 
   base_properties::update_handlevisibility ();
 }
+
+static void
+update_text_pos (graphics_handle h)
+{
+   graphics_object go = gh_manager::get_object (h);
+   if (go.isa ("text"))
+   {
+      text::properties& tp
+      = dynamic_cast<text::properties&> (go.get_properties ());
+      tp.update_font ();
+      tp.update_text_extent ();
+   }
+   else if (go.isa ("figure") || go.isa ("uipanel") || go.isa ("axes")
+            || go.isa ("hggroup"))
+   {
+      Matrix ch = go.get_properties ().get_all_children ();
+      for (octave_idx_type ii = 0; ii < ch.numel (); ii++)
+         update_text_pos (graphics_handle (ch(ii)));
+      
+      if (go.isa ("axes"))
+      {
+         axes::properties& ap
+         = dynamic_cast<axes::properties&> (go.get_properties ());
+         ap.update_font ();
+         ap.update_xlabel_position ();
+         ap.update_ylabel_position ();
+         ap.update_zlabel_position ();
+         ap.update_title_position ();
+         ap.update_axes_layout ();
+      }
+   }
+}
+
+void
+figure::properties::update___device_pixel_ratio__ (void)
+{
+   update_text_pos (get___myhandle__ ());
+}
+
 // ---------------------------------------------------------------------
 
 void
@@ -6581,11 +6634,13 @@
 
     }
 
+   double dpr = device_pixel_ratio (get___myhandle__ ());
+   
   gh_manager::auto_lock guard;
   txt_renderer.set_font (get ("fontname").string_value (),
                          get ("fontweight").string_value (),
                          get ("fontangle").string_value (),
-                         get ("__fontsize_points__").double_value ());
+                         get ("__fontsize_points__").double_value () * dpr);
 }
 
 // The INTERNAL flag defines whether position or outerposition is used.
@@ -7654,6 +7709,7 @@
 {
   Matrix ext (1, 2, 0.0);
   double wmax, hmax;
+   double dpr = device_pixel_ratio (get___myhandle__ ());
   wmax = hmax = 0.0;
   int n = std::min (ticklabels.numel (), ticks.numel ());
   for (int i = 0; i < n; i++)
@@ -7671,8 +7727,8 @@
               ext = txt_renderer.get_extent (label, 0.0,
                                              get_ticklabelinterpreter ());
 
-              wmax = std::max (wmax, ext(0));
-              hmax = std::max (hmax, ext(1));
+              wmax = std::max (wmax, ext(0) / dpr);
+              hmax = std::max (hmax, ext(1) / dpr);
             }
           else
             {
@@ -8699,7 +8755,14 @@
   m(0) += p(0);
   m(1) += p(1);
 
-  return convert_text_position (m, *this, "pixels", get_units ());
+   Matrix bbox = convert_text_position (m, *this, "pixels", get_units ());
+   
+   double dpr = device_pixel_ratio (get___myhandle__ ());
+   
+   for (octave_idx_type ii = 0; ii < bbox.numel (); ii++)
+      bbox(ii) = bbox(ii) / dpr;
+   
+   return bbox;
 }
 
 void
@@ -8737,11 +8800,13 @@
 void
 text::properties::update_font (void)
 {
+   double dpr = device_pixel_ratio (get___myhandle__ ());
+   
   gh_manager::auto_lock guard;
   txt_renderer.set_font (get ("fontname").string_value (),
                          get ("fontweight").string_value (),
                          get ("fontangle").string_value (),
-                         get ("__fontsize_points__").double_value ());
+                         get ("__fontsize_points__").double_value () * dpr);
 
   txt_renderer.set_color (get_color_rgb ());
 }
@@ -8838,9 +8903,9 @@
   double fontsz = get_fontsize ();
   double parent_height = box_pix_height;
 
+   graphics_object go (gh_manager::get_object (get___myhandle__ ()));
   if (fontunits_is ("normalized") && parent_height <= 0)
     {
-      graphics_object go (gh_manager::get_object (get___myhandle__ ()));
       graphics_object ax (go.get_ancestor ("axes"));
 
       parent_height = ax.get_properties ().get_boundingbox (true).elem (3);
diff -r 1f46d371968c libinterp/corefcn/graphics.in.h
--- a/libinterp/corefcn/graphics.in.h	Thu Aug 09 14:20:32 2018 -0400
+++ b/libinterp/corefcn/graphics.in.h	Tue Sep 24 16:22:28 2019 +0200
@@ -3178,6 +3178,7 @@
       any_property __plot_stream__ h , Matrix ()
       any_property __rotate_mode__ h , Matrix ()
       any_property __zoom_mode__ h , Matrix ()
+      double_property __device_pixel_ratio__ hU , 1.0
 
       // Obsolete properties: doublebuffer, mincolormap, wvisual, wvisualmode,
       //                      xdisplay, xvisual, xvisualmode
@@ -3521,7 +3522,7 @@
     void clear_zoom_stack (bool do_unzoom = true);
 
     void update_units (const caseless_str& old_units);
-
+     void update_font (std::string prop = "");
     void update_fontunits (const caseless_str& old_fontunits);
 
   private:
@@ -3956,7 +3957,6 @@
         calc_ticklabels (ztick, zticklabel, zscale.is ("log"), false, 2, zlim);
     }
 
-    void update_font (std::string prop = "");
     void update_fontname (void)
     {
       update_font ("fontname");
@@ -4385,6 +4385,10 @@
   public:
     double get___fontsize_points__ (double box_pix_height = 0) const;
 
+     void update_text_extent (void);
+     
+     void update_font (void);
+     
     void set_position (const octave_value& val)
     {
       octave_value new_val (val);
@@ -4494,15 +4498,12 @@
         set_zliminclude ("off");
     }
 
-    void update_text_extent (void);
-
     void request_autopos (void);
     void update_positionmode (void) { request_autopos (); }
     void update_rotationmode (void) { request_autopos (); }
     void update_horizontalalignmentmode (void) { request_autopos (); }
     void update_verticalalignmentmode (void) { request_autopos (); }
 
-    void update_font (void);
     void update_string (void) { request_autopos (); update_text_extent (); }
     void update_rotation (void) { update_text_extent (); }
     void update_color (void) { update_font (); update_text_extent (); }
diff -r 1f46d371968c m4/acinclude.m4
--- a/m4/acinclude.m4	Thu Aug 09 14:20:32 2018 -0400
+++ b/m4/acinclude.m4	Tue Sep 24 16:22:28 2019 +0200
@@ -666,6 +666,36 @@
   fi
 ])
 dnl
+dnl Check whether the Qt class QScreen has the devicePixelRatio member function.
+dnl This member function was introduced in Qt 5.5.
+dnl
+AC_DEFUN([OCTAVE_CHECK_FUNC_QSCREEN_DEVICEPIXELRATIO], [
+  AC_CACHE_CHECK([for QScreen::devicePixelRatio in <QScreen>],
+    [octave_cv_func_qscreen_devicepixelratio],
+    [AC_LANG_PUSH(C++)
+    ac_octave_save_CPPFLAGS="$CPPFLAGS"
+    ac_octave_save_CXXFLAGS="$CXXFLAGS"
+    CPPFLAGS="$QT_CPPFLAGS $CXXPICFLAG $CPPFLAGS"
+    CXXFLAGS="$CXXPICFLAG $CXXFLAGS"
+    AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
+        #include <QApplication>
+        #include <QScreen>
+        ]], [[
+        QScreen *screen = QApplication::primaryScreen ();
+        qreal ratio = screen->devicePixelRatio ();
+        ]])],
+      octave_cv_func_qscreen_devicepixelratio=yes,
+      octave_cv_func_qscreen_devicepixelratio=no)
+    CPPFLAGS="$ac_octave_save_CPPFLAGS"
+    CXXFLAGS="$ac_octave_save_CXXFLAGS"
+    AC_LANG_POP(C++)
+  ])
+  if test $octave_cv_func_qscreen_devicepixelratio = yes; then
+    AC_DEFINE(HAVE_QSCREEN_DEVICEPIXELRATIO, 1,
+      [Define to 1 if you have the `QScreen::devicePixelRatio' member function.])
+  fi
+])
+dnl
 dnl Check whether the Qt class QTabWidget has the setMovable member function.
 dnl This member function was introduced in Qt 4.5.
 dnl
@@ -2107,6 +2137,7 @@
     OCTAVE_CHECK_FUNC_QLINEEDIT_SETPLACEHOLDERTEXT
     OCTAVE_CHECK_FUNC_QMOUSEEVENT_LOCALPOS
     OCTAVE_CHECK_FUNC_QOBJECT_FINDCHILDREN_ACCEPTS_FINDCHILDOPTIONS
+    OCTAVE_CHECK_FUNC_QSCREEN_DEVICEPIXELRATIO
     OCTAVE_CHECK_FUNC_QTABWIDGET_SETMOVABLE
     OCTAVE_CHECK_FUNC_QTMESSAGEHANDLER_ACCEPTS_QMESSAGELOGCONTEXT
     OCTAVE_CHECK_MEMBER_QFONT_FORCE_INTEGER_METRICS
diff -r 1f46d371968c scripts/plot/util/print.m
--- a/scripts/plot/util/print.m	Thu Aug 09 14:20:32 2018 -0400
+++ b/scripts/plot/util/print.m	Tue Sep 24 16:22:28 2019 +0200
@@ -382,6 +382,15 @@
       nfig += 1;
     endfor
 
+    ## Don't account for the actual pixel density
+    if (strcmp (tk, "qt"))
+      props(end+1).h = opts.figure;
+      props(end).name = "__device_pixel_ratio__";
+      props(end).value = {get(opts.figure, "__device_pixel_ratio__")};
+      set (opts.figure, "__device_pixel_ratio__", 1);
+      nfig += 1;
+    endif
+
     ## print() requires axes units = "normalized"
     hax = findall (opts.figure, "-depth", 1, "type", "axes", ...
       "-not", "units", "normalized");
